### PR TITLE
[move] Add save hooks for extracting compiler information 

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

- Most of the usages of the stepped compiler are to simply clone out portions of the AST
  - Particularly for the move-model and move-model-2
- I added generalized hooks that let you save any step of compilation, including ProgramInfo, which might be helpful for the IDE

## Test plan 

- Internal change

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/command_line/compiler.rs [hunk 1]
```diff
@@ -5,18 +5,18 @@
...
-    compiled_unit,
-    compiled_unit::AnnotatedCompiledUnit,
+    compiled_unit::{self, AnnotatedCompiledUnit},
     diagnostics::{
         codes::{Severity, WarningFilter},
         *,
     },
     editions::Edition,
-    expansion, hlir, interface_generator, naming, parser,
-    parser::{comments::*, *},
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/command_line/compiler.rs:16]: `self` plus `*` seems dubious?

we should really write a style guide for imports and make the code all use it...

**Comment by cgswords** [comment]
Can we simplify this implementation quite a bit?

```
pub struct SavedInfo {
    parser: Option<P::Program>,
    expansion: Option<E::Program>,
    naming: Option<N::Program>,
    typing: Option<T::Program>,
    typing_info: Option<Arc<program_info::TypingProgramInfo>>,
    hlir: Option<H::Program>,
    cfgir: Option<G::Program>,
}

#[derive(Ord, PartialOrd)]
pub enum SavedInfoFlag {
    Parser,
    Expansion,
    Naming,
    Typing,
    TypingInfo,
    HLIR,
    CFGIR,
}

pub type SavedInfoSettings = BTreeSet<SavedInfoFlag>;

...

pub struct CompilationEnv {
    ...,
    saved_info: Rc<RefCell<SavedInfo>>,
    save_info_flags: SavedInfoSettings,
}

...

impl CompilationEnv {

    ...

    pub fn save_parser_ast(&self, ast: &P::Program) {
        if self.save_info_flags.contains(SavedInfoFlag::Parser) {
            self.saved_info.borrow_mut().parser = Some(ast.clone());
        }
    }

    ...

    pub fn set_saved_flag(&mut self, flag: SavedInfoFlag) {
        self.saved_info_flags.insert(flag);
    }
}
  ```

File: external-crates/move/crates/move-compiler/src/command_line/compiler.rs [hunk 2]
```diff
@@ -62,6 +62,8 @@ pub struct Compiler {
     default_config: Option<PackageConfig>,
     /// Root path of the virtual file system.
     vfs_root: Option<VfsPath>,
+    /// Hooks to save the ASTs
+    save_hooks: Vec<SaveHook>,
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/command_line/compiler.rs:66]: Is there a strong motivation to keep these as a vector?

**Comment by tnowacki** [comment]
> Can we simplify this implementation quite a bit?

I like it. Will do

**Review by cgswords — APPROVED** [review]
Awesome!

**Pull request merged**